tests/ReferenceTest.php

```diff
@@ -0,0 +1,305 @@
+                    '$a===' => '2',
+                    '$b===' => '3',
+                    '$c===' => '2',
+                    '$d===' => '3',
+                ],
+            ],
+            'referenceToSubsequentCatch' => [
+                '<?php
+                    $a = null;
+                    $b = &$a;
+
+                    try {
+                        throw new \Exception();
+                    } catch (\Exception $a) {
+                        takesException($b);
+                    }
+                    function takesException(\Exception $e): void {}
+                ',
+            ],
+            'referenceAsSubsequentCatch' => [
+                '<?php
+                    $a = null;
+                    $b = &$a;
+
+                    try {
+                        throw new \Exception();
+                    } catch (\Exception $b) {
+                        takesException($a);
+                    }
+                    function takesException(\Exception $e): void {}
+                ',
+            ],
+            'referenceToNewVariableInitializesNull' => [
+                '<?php
+                    $b = &$a;
+                ',
+                'assertions' => [
+                    '$a===' => 'null',
+                    '$b===' => 'null',
+                ],
+            ],
+            'referenceShadowedByGlobal' => [
+                '<?php
+                    /** @var string */
+                    $a = 0;
+                    function foo(): void
+                    {
+                        $b = 1;
+                        $a = &$b;
+                        global $a;
+                        takesString($a);
+                    }
+
+                    function takesString(string $str): void {}
+                ',
+            ],
+            'unsetPreventsReferenceConfusion' => [
+                '<?php
+                    $arr = [1, 2, 3];
+                    foreach ($arr as &$i) {
+                        ++$i;
+                    }
+                    unset($i);
+
+                    for ($i = 0; $i < 10; ++$i) {
+                        echo $i;
+                    }
+                ',
+            ],
+            'assignmentAsReferencePreventsReferenceConfusion' => [
+                '<?php
+                    $arr = [1, 2, 3];
+                    foreach ($arr as &$i) {
+                        ++$i;
+                    }
+
+                    $i = &$foo;
+
+                    for ($i = 0; $i < 10; ++$i) {
+                        echo $i;
+                    }
+                ',
+            ],
+            'assignmentAsReferenceInForeachPreventsReferenceConfusion' => [
+                '<?php
+                    $arr = [1, 2, 3];
+                    foreach ($arr as &$i) {
+                        ++$i;
+                    }
+                    foreach ($arr as &$i) {
+                        ++$i;
+                    }
+                ',
+            ],
+        ];
```

Comment:
can you add a test to see if https://3v4l.org/omrIG makes Psalm crashes and/or create a blackhole?

Reply:
Also this kind of mumbo jumbo: https://3v4l.org/hsBCq